libsolidity/ast/Types.cpp

```diff
@@ -689,31 +754,33 @@ tuple<bool, rational> RationalNumberType::isValidLiteral(Literal const& _literal
 		}
 		else if (expPoint != _literal.value().end())
 		{
-			// parse the exponent
+			// Parse base and exponent. Checks numeric limit.
 			bigint exp = bigint(string(expPoint + 1, _literal.value().end()));
-
-			if (exp > numeric_limits<int32_t>::max() || exp < numeric_limits<int32_t>::min())
-				return make_tuple(false, rational(0));
-
-			// parse the base
 			tuple<bool, rational> base = parseRational(string(_literal.value().begin(), expPoint));
+
 			if (!get<0>(base))
 				return make_tuple(false, rational(0));
 			value = get<1>(base);
 
 			if (exp < 0)
 			{
 				exp *= -1;
+				if (!fitsPrecisionBase10(abs(value.denominator()), exp))
+					return make_tuple(false, rational(0));
 				value /= boost::multiprecision::pow(
```

Comment:
add assert (or check) that `exp > numeric_limits<int32_t>::max() || exp < numeric_limits<int32_t>::min()`